File: projects/slick/lib/slick/call_handler.rb

require "slick/helpers"

class Slick::CallHandler

    include Slick::Helpers

    def handle_call
        view_name = request.path.sub(/\A\//, '')
    
        begin
            render_guard_views(view_name)
            return if response.body.length > 0
            if !view_name.match(/(\A|\/)_/)
                view_name_segments = view_name.split(/\//)
                candidate_index_view_name = (view_name_segments + ['index']).join('/');
                render_view(candidate_index_view_name, params) if view?(candidate_index_view_name)
                return if response.body.length > 0
                render_view(view_name, params) if view?(view_name)
                return if response.body.length > 0
            end
            render_default_views(view_name)
        rescue Slick::ExitException => e
            Slick::Workspace.new.instance_eval(&e.block) if e.block
        end

        if response.body.length == 0
            response.status = 404
            response.set_header("Content-Type", "text/plain")
            response.body = ["Not found"]
        end
    end

    def render_guard_views(view_name)
        view_name_segments = view_name.split(/\//)

        candidate_index_view_name = (view_name_segments + ['index']).join('/');
        candidate_default_view_name = (view_name_segments + ['default']).join('/');

        view_name_segments.pop if !view?(candidate_index_view_name) && !view?(candidate_default_view_name)

        prefix_segments = [];
        while true
            candidate_guard_view_name = (prefix_segments + ['guard']).join('/')
            render_view(candidate_guard_view_name, params) if view?(candidate_guard_view_name)
            break if response.body.length > 0 || view_name_segments.length == 0
            prefix_segments.push(view_name_segments.shift)
        end
    end

    def render_default_views(view_name)
        prefix_segments = view_name.split(/\//)
        while true
            candidate_default_view_name = (prefix_segments + ['default']).join('/')
            render_view(candidate_default_view_name, params) if view?(candidate_default_view_name)
            break if response.body.length > 0 || break if prefix_segments.length == 0
            prefix_segments.pop
        end
    end

    def view?(name)
        Slick::View.registered_classes[name] != nil
    end

end

File: packages/pinstripe/lib/services/call_handler.js

import { Workspace } from "../workspace.js";

export default {
    meta(){
        this.addToClient();
    },

    create(){
        return this;
    },
    
    async handleCall(params = {}, useContext = false){
        if(!useContext) return Workspace.run(function(){
            return this.callHandler.handleCall(params, true);
        });

        this.context.params = this.normalizeParams(params);
        
        const out = this.normalizeResponse(await this.render());
        if(out) return out;

        return [404, {'content-type': 'text/plain'}, ['Not found']];
    },

    async render(){
        const viewName = this.params._url.pathname.replace(/^\/|\/$/g, '') || 'index';
        
        let out = await this.renderGuardViews(viewName, this.params);
        if(out) return out;

        if(!viewName.match(/(^|\/)_[^\/]+(|\/index)$/)){
            out = await this.renderView(viewName, this.params);
            if(out) return out;
        }
        
        return this.renderDefaultViews(viewName, this.params);
    },

    async renderGuardViews(viewName, params){
        const viewNameSegments = viewName != '' ? viewName.split(/\//) : [];

        const prefixSegments = [];
        while(true){
            const candidateGuardViewName = prefixSegments.length ? [...prefixSegments, 'guard'].join('/') : 'guard';
            const out = await this.renderView(candidateGuardViewName, params);
            if(out) return out;
            if(viewNameSegments.length == 0) break;
            prefixSegments.push(viewNameSegments.shift());
        }
    },

    async renderDefaultViews(viewName, params){
        const prefixSegments = viewName != '' ? viewName.split(/\//) : [];
        while(true){
            const candidateDefaultViewName = prefixSegments.length ? [...prefixSegments, 'default'].join('/') : 'default';
            const out = await this.renderView(candidateDefaultViewName, params);
            if(out) return out;
            if(prefixSegments.length == 0) break;
            prefixSegments.pop();
        }
    },

    normalizeParams(params){
        const out = { ...params }
        if(!out._method) out._method = 'get';
        out._url = new URL(out._url ?? '/', 'http://localhost');
        if(!params._headers) out._headers = {};
        return out;
    },

    normalizeResponse(response){
        if(!response) return;

        let responseArray;
        if(response && typeof response.toResponseArray == 'function'){
            responseArray = response.toResponseArray();
        } else if(response && !Array.isArray(response)){
            responseArray = [200, {'content-type': 'text/plain'}, [`${response}`]];
        } else {
            responseArray = response;
        }

        const [ status, headers, body ] = responseArray;
        const normalizedHeaders = {};
        Object.keys(headers).forEach(name => {
            normalizedHeaders[name.toLowerCase()] = headers[name];
        });

        return [ status, normalizedHeaders, body ];
    }
};


